core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,424 @@
+pub(super) struct BlockStore {
+    inner: Store,
+    cursor: sync::Mutex<Option<Cursor>>,
+}
+
+impl Store {
+    /// Creates a `Store`. `pool` should have multiple connections to work efficiently.
+    pub fn new(pool: ConnectionPool, operator_address: Address) -> Self {
+        Self {
+            pool,
+            operator_address,
+        }
+    }
+
+    /// Converts `Store` into a `BlockStore`.
+    pub fn into_block_store(self) -> BlockStore {
+        BlockStore {
+            inner: self,
+            cursor: sync::Mutex::new(None),
+        }
+    }
+}
+
+impl BlockStore {
+    /// Generates and stores the genesis cert (signed by `validator_key`) for the last sealed miniblock.
+    /// No-op if db already contains a genesis cert.
+    pub async fn try_init_genesis(
+        &mut self,
+        ctx: &ctx::Ctx,
+        validator_key: &validator::SecretKey,
+    ) -> ctx::Result<()> {
+        let mut storage = CtxStorage::access(ctx, &self.inner.pool)
+            .await
+            .wrap("access()")?;
+        // Fetch last miniblock number outside of the transaction to avoid taking a lock.
+        let number = storage
+            .last_miniblock_number(ctx)
+            .await
+            .wrap("last_miniblock_number()")?;
+
+        let mut txn = storage
+            .start_transaction(ctx)
+            .await
+            .wrap("start_transaction()")?;
+        if txn
+            .first_certificate(ctx)
+            .await
+            .wrap("first_certificate()")?
+            .is_some()
+        {
+            return Ok(());
+        }
+        let payload = txn
+            .payload(ctx, number, self.inner.operator_address)
+            .await
+            .wrap("payload()")?
+            .context("miniblock disappeared")?;
+        let (genesis, _) = zksync_consensus_bft::testonly::make_genesis(
+            &[validator_key.clone()],
+            payload.encode(),
+            number,
+        );
+        txn.insert_certificate(ctx, &genesis.justification, self.inner.operator_address)
+            .await
+            .wrap("insert_certificate()")?;
+        txn.commit(ctx).await.wrap("commit()")
+    }
+
+    /// Sets an `ActionQueueSender` in the `BlockStore`. See `store_next_block()` for details.
+    pub async fn set_actions_queue(
+        &mut self,
+        ctx: &ctx::Ctx,
+        actions: ActionQueueSender,
+    ) -> ctx::Result<()> {
+        let mut storage = CtxStorage::access(ctx, &self.inner.pool)
+            .await
+            .wrap("access()")?;
+        let inner = storage
+            .new_fetcher_cursor(ctx)
+            .await
+            .wrap("new_fetcher_cursor()")?;
+        *sync::lock(ctx, &self.cursor).await? = Some(Cursor { inner, actions });
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl PersistentBlockStore for BlockStore {
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<BlockStoreState> {
+        // Ensure that genesis block has consensus field set in postgres.
```

Comment:
Is this comment outdated? Also, how *is* the genesis block set for ENs? Or is it not ensured specifically? IIUC, in tests it's currently obtained from a Postgres-level snapshot. Is that the intended way to pass the genesis block data for ENs?

Reply:
Comment removed. Yes, ENs are supposed to obtain the certificate via the snapshot for simplicity.